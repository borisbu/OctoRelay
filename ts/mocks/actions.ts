export const cancelMock = jest.fn();

jest.mock("../helpers/actions", () => ({
  cancelTask: cancelMock,
}));
