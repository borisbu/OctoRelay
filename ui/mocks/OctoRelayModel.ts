import { vi } from "vitest";

export const modelMock = vi.fn();

vi.mock("../model/OctoRelayModel", () => ({ OctoRelayViewModel: modelMock }));
