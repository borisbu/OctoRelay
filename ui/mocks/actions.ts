export const cancelMock = jest.fn();
export const toggleMock = jest.fn();

jest.mock("../helpers/actions", () => ({
  cancelTask: cancelMock,
  toggleRelay: toggleMock,
}));
