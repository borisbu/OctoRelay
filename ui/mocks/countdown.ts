import { vi } from "vitest";

export const disposerMock = vi.fn();
export const countdownMock = vi.fn(() => disposerMock);
export const deadlineMock = vi.fn(() => "sample deadline");

vi.mock("../helpers/countdown", () => ({
  setCountdown: countdownMock,
  formatDeadline: deadlineMock,
}));
