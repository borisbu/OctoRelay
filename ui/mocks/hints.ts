import { vi } from "vitest";

export const clearMock = vi.fn();
export const showMock = vi.fn();

vi.mock("../helpers/hints", () => ({
  clearHints: clearMock,
  showHints: showMock,
}));
