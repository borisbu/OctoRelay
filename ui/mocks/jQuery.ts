export const elementMock: Record<
  | "toggle"
  | "html"
  | "off"
  | "on"
  | "find"
  | "text"
  | "modal"
  | "is"
  | "popover"
  | "tooltip",
  jest.Mock
> = {
  toggle: jest.fn(() => elementMock),
  html: jest.fn(() => elementMock),
  off: jest.fn(() => elementMock),
  on: jest.fn(() => elementMock),
  find: jest.fn(() => elementMock),
  text: jest.fn(() => elementMock),
  modal: jest.fn(() => elementMock),
  is: jest.fn(() => elementMock),
  tooltip: jest.fn(() => elementMock),
  popover: jest.fn(() => elementMock),
};

export const jQueryMock = jest.fn((subject: string | (() => void)) => {
  if (typeof subject === "function") {
    return subject();
  }
  return elementMock;
});
