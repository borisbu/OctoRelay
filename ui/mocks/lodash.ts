import { vi } from "vitest";

export const lodashMock = {
  memoize: vi.fn((fn: () => unknown) => fn),
};
