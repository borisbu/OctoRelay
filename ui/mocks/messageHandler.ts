import { vi } from "vitest";

export const handlerMock = vi.fn();

vi.mock("../model/messageHandler", () => ({
  makeMessageHandler: vi.fn(() => handlerMock),
}));
