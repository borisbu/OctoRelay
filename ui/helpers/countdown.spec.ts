import MockDate from "mockdate";
import assert from "node:assert/strict";
import { elementMock, jQueryMock } from "../mocks/jQuery";
import { lodashMock } from "../mocks/lodash";

describe("Countdown helpers", async () => {
  const warnSpy = vi.spyOn(console, "warn").mockImplementation(() => {});
  const setIntervalMock = vi.fn<(handler: () => void, delay: number) => void>(
    () => "mockedInterval",
  );
  const clearIntervalMock = vi.fn();

  Object.assign(global, {
    LOCALE: "en",
    $: jQueryMock,
    _: lodashMock,
    setInterval: setIntervalMock,
    clearInterval: clearIntervalMock,
  });

  const { formatDeadline, getCountdownDelay, setCountdown } = await import(
    "./countdown"
  );

  beforeAll(() => {
    MockDate.set("2023-08-13T22:30:00");
  });

  afterEach(() => {
    Object.assign(global, {
      LOCALE: "en",
    });
    warnSpy.mockClear();
    setIntervalMock.mockClear();
    clearIntervalMock.mockClear();
    elementMock.text.mockClear();
    elementMock.is.mockClear();
  });

  afterAll(() => {
    MockDate.reset();
  });

  describe("formatDeadline() helper", () => {
    test.each([0, 1, 30, 59, 60, 61, 90, 119, 120, 121, 300, 600, 3600])(
      "Should format the supplied UNIX timestamp having offset %s seconds",
      (offset) => {
        expect(formatDeadline(Date.now() + offset * 1000)).toMatchSnapshot();
        expect(warnSpy).not.toHaveBeenCalled();
      },
    );

    test("should handle invalid locales", () => {
      Object.assign(global, {
        LOCALE: "invalid_locale_id",
      });
      expect(formatDeadline(Date.now() + 10000)).toBe(`in 10 seconds`);
      expect(warnSpy).toHaveBeenCalledTimes(2);
      expect(warnSpy.mock.calls).toEqual([
        [
          "Failed to format time using invalid_locale_id locale",
          expect.any(Error),
        ],
        [
          "Failed to format time using invalid-locale-id locale",
          expect.any(Error),
        ],
      ]);
    });

    test.each([
      [1000, "1 second"],
      [10000, "10 seconds"],
    ])("should handle complete Intl malfunction", (offset, expected) => {
      vi.spyOn(Intl, "NumberFormat").mockImplementation(() =>
        assert.fail("Can not do this"),
      );
      expect(formatDeadline(Date.now() + offset)).toBe(`in ${expected}`);
      expect(warnSpy).toHaveBeenCalledTimes(2);
      expect(warnSpy.mock.calls).toEqual([
        ["Failed to format time using en locale", expect.any(Error)],
        ["Failed to format time using undefined locale", expect.any(Error)],
      ]);
    });
  });

  describe("getCountdownDelay() helper", () => {
    test.each([60, 120, 121])(
      "should return refresh interval for supplied timestamp having offset %s seconds",
      (offset) => {
        expect(getCountdownDelay(Date.now() + offset * 1000)).toMatchSnapshot();
      },
    );
  });

  describe("setCountdown() helper", () => {
    test.each([true, false])("should set countdown %#", (isVisible) => {
      setCountdown(elementMock as unknown as JQuery, Date.now() + 121 * 1000);
      expect(setIntervalMock).toHaveBeenCalledWith(expect.any(Function), 60000);
      const intervalFn = setIntervalMock.mock.calls[0][0];
      elementMock.is.mockImplementationOnce(() => isVisible);
      MockDate.set(Date.now() + 1000); // this will trigger the new delay
      intervalFn();
      if (isVisible) {
        expect(elementMock.text).toHaveBeenCalledWith("in 2 minutes");
        expect(setIntervalMock).toHaveBeenCalledTimes(2); // reset with a new delay
        expect(setIntervalMock).toHaveBeenCalledWith(
          expect.any(Function),
          1000,
        );
        // coverage branch for the case nextDelay === delay
        const nextIntervalFn = setIntervalMock.mock.calls[1][0];
        MockDate.set(Date.now() + 1000); // same delay branch this time
        nextIntervalFn();
        expect(setIntervalMock).toHaveBeenCalledTimes(2); // not called again this time
      } else {
        expect(elementMock.text).not.toHaveBeenCalled();
        expect(clearIntervalMock).toHaveBeenCalledWith("mockedInterval");
      }
    });
  });
});
