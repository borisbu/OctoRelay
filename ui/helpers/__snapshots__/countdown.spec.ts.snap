// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 0 seconds 1`] = `"in 0 seconds"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 1 seconds 1`] = `"in 1 second"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 30 seconds 1`] = `"in 30 seconds"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 59 seconds 1`] = `"in 59 seconds"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 60 seconds 1`] = `"in 1.0 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 61 seconds 1`] = `"in 1.0 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 90 seconds 1`] = `"in 1.5 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 119 seconds 1`] = `"in 2.0 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 120 seconds 1`] = `"in 2 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 121 seconds 1`] = `"in 2 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 300 seconds 1`] = `"in 5 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 600 seconds 1`] = `"in 10 minutes"`;

exports[`Countdown helpers formatDeadline() helper Should format the supplied UNIX timestamp having offset 3600 seconds 1`] = `"in 1 hour"`;

exports[`Countdown helpers getCountdownDelay() helper should return refresh interval for supplied timestamp having offset 60 seconds 1`] = `1000`;

exports[`Countdown helpers getCountdownDelay() helper should return refresh interval for supplied timestamp having offset 120 seconds 1`] = `1000`;

exports[`Countdown helpers getCountdownDelay() helper should return refresh interval for supplied timestamp having offset 121 seconds 1`] = `60000`;
